beater/api/mux.go

```diff
@@ -155,29 +133,49 @@ func sourcemapHandler(cfg *config.Config, reporter publish.Reporter) (request.Ha
 	return middleware.Wrap(h, sourcemapMiddleware(cfg)...)
 }
 
-func sourcemapMiddleware(cfg *config.Config) []middleware.Middleware {
-	return append(apmMiddleware(sourcemap.MonitoringMap),
-		middleware.KillSwitchMiddleware(cfg.RumConfig.IsEnabled() && cfg.RumConfig.SourceMapping.IsEnabled()),
-		middleware.RequireAuthorizationMiddleware(cfg.SecretToken))
+func backendAgentConfigHandler(cfg *config.Config, _ publish.Reporter) (request.Handler, error) {
+	return agentConfigHandler(cfg, backendMiddleware(cfg, agent.MonitoringMap))
 }
 
-func configAgentHandler(cfg *config.Config, _ publish.Reporter) (request.Handler, error) {
+func agentConfigHandler(cfg *config.Config, m []middleware.Middleware) (request.Handler, error) {
 	var kbClient kibana.Client
 	if cfg.Kibana.Enabled() {
 		kbClient = kibana.NewConnectingClient(cfg.Kibana)
 	}
 	h := agent.Handler(kbClient, cfg.AgentConfig)
-	return middleware.Wrap(h, agentConfigMiddleware(cfg)...)
+	ks := middleware.KillSwitchMiddleware(cfg.Kibana.Enabled())
+	return middleware.Wrap(h, append(m, ks)...)
 }
 
-func agentConfigMiddleware(cfg *config.Config) []middleware.Middleware {
-	return append(apmMiddleware(agent.MonitoringMap),
-		middleware.KillSwitchMiddleware(cfg.Kibana.Enabled()),
+func rootHandler(cfg *config.Config, _ publish.Reporter) (request.Handler, error) {
+	return middleware.Wrap(root.Handler(), rootMiddleware(cfg)...)
+}
+
+func apmMiddleware(m map[request.ResultID]*monitoring.Int) []middleware.Middleware {
+	return []middleware.Middleware{
+		middleware.LogMiddleware(),
+		middleware.RecoverPanicMiddleware(),
+		middleware.MonitoringMiddleware(m),
+		middleware.RequestTimeMiddleware(),
+	}
+}
+
+func backendMiddleware(cfg *config.Config, m map[request.ResultID]*monitoring.Int) []middleware.Middleware {
+	return append(apmMiddleware(m),
 		middleware.RequireAuthorizationMiddleware(cfg.SecretToken))
 }
 
-func rootHandler(cfg *config.Config, _ publish.Reporter) (request.Handler, error) {
-	return middleware.Wrap(root.Handler(), rootMiddleware(cfg)...)
+func rumMiddleware(cfg *config.Config, m map[request.ResultID]*monitoring.Int) []middleware.Middleware {
+	return append(apmMiddleware(m),
+		middleware.SetRateLimitMiddleware(cfg.RumConfig.EventRate),
+		middleware.CORSMiddleware(cfg.RumConfig.AllowOrigins),
+		middleware.KillSwitchMiddleware(cfg.RumConfig.IsEnabled()))
```

Comment:
Actually IMO the `KillSwitchHandler` should be the first middleware to run. Why do you run it at the end and waste resources when the endpoint is disabled?

Reply:
Mostly because of convenience, and because it is easy to be consistent about it. I don't expect anyone hammering a disabled endpoint and wasting resources, more likely someone will notice it after a one or few queries and enable it.

Reply:
> Mostly because of convenience, and because it is easy to be consistent about it. 

I don't see the inconvenience or inconsistency if the order is changed. 

In case of an attack it would waste resources though, so I suggest to change the order.